core/server/master/src/main/java/alluxio/master/backup/BackupWorkerRole.java

```diff
@@ -0,0 +1,378 @@
+  }
+
+  /**
+   * Handler for suspend message. It's used in secondary master.
+   */
+  private void handleSuspendJournalsMessage(BackupSuspendMessage suspendMsg) {
+    LOG.info("Received suspend message: {}", suspendMsg.toString());
+    Preconditions.checkState(!mBackupTracker.inProgress(), "Backup in progress");
+
+    try {
+      mJournalSystem.suspend();
+      LOG.info("Suspended journals for backup.");
+    } catch (IOException e) {
+      String failMessage = "Failed to suspended journals for backup.";
+      LOG.error(failMessage, e);
+      throw new RuntimeException(failMessage, e);
+    }
+    // Schedule a timeout task to resume journals if protocol is not followed by the leader.
+    mBackupTimeoutTask = mTaskScheduler.schedule(() -> {
+      LOG.info("Resuming journals as backup request hasn't been received.");
+      enforceResumeJournals();
+    }, BACKUP_ABORT_AFTER_SUSPEND_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Handler for backup request message. It's used in secondary master.
+   */
+  private void handleRequestMessage(BackupRequestMessage requestMsg) {
+    LOG.info("Received backup message: {}", requestMsg.toString());
+    Preconditions.checkState(!mBackupTracker.inProgress(), "Backup in progress");
+
+    // Reset backup tracker.
+    mBackupTracker.reset();
+    // Update current backup status with given backup id.
+    mBackupTracker.update(new BackupStatus(requestMsg.getBackupId(), BackupState.Initiating));
+    mBackupTracker.updateHostname(NetworkAddressUtils.getLocalHostName(
+        (int) ServerConfiguration.global().getMs(PropertyKey.NETWORK_HOST_RESOLUTION_TIMEOUT_MS)));
+
+    // Start sending backup progress to leader.
+    startHeartbeatThread();
+
+    // Cancel timeout task created by suspend message handler.
+    if (!mBackupTimeoutTask.cancel(true)) {
+      mBackupTracker.updateError(new BackupException("Journal has ben resumed due to a time-out"));
+      return;
+    }
+
+    // Spawn a task for advancing journals to target sequences, then taking the backup.
+    mBackupFuture = mExecutorService.submit(() -> {
+      // Mark state as transitioning.
+      mBackupTracker.updateState(BackupState.Transitioning);
+      try {
+        LOG.info(
+            "Initiating catching up of journals to consistent sequences before starting backup. {}",
+            requestMsg.getJournalSequences());
+        CatchupFuture catchupFuture = mJournalSystem.catchup(requestMsg.getJournalSequences());
+        CompletableFuture.runAsync(() -> catchupFuture.waitTermination())
+            .get(BACKUP_ABORT_AFTER_TRANSITION_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+      } catch (Exception e) {
+        mBackupTracker.updateError(new BackupException(
+            String.format("Failed to catch-up journals: %s", e.getMessage()), e));
+        // Don't resume journals if interrupted.
+        // Catch-up future can be interrupted only when master becomes primary during backup.
```

Comment:
Is this true, don't we cancel the future if we lose connection to the master?

Reply:
Nice catch!

This statement is not correct:

First; the future that is being interrupted is not the `catchup-future` but the async task that is spawned to wait for `catchup-future`. (Waiting on catchup-future is not interruptable)
Second; we interrupt this code when connection to master is lost.

The reason, I didn't enforce resuming when interrupted was to quit from stopMasters() path ASAP during gain-primacy scenario. However the journal itself will still call it when gaining primacy, so being quick here does not really have a value. So removing this code completely.